File: Visual Basic.Net/Ginkgo I2C/VB.NET_USB_I2C_AT24C02/USB_I2C_AT24C02/USB_I2C_AT24C02/ControlI2C.vb
Option Strict Off
Option Explicit On

Imports System.Runtime.InteropServices
Module ControlI2C
    'General Error Code
    Public Const SUCCESS As Int32 = 0                   ' no Error
    Public Const PARAMETER_NULL As Int32 = -1           ' void pointer 
    Public Const INPUT_DATA_TOO_MUCH As Int32 = -2      ' too many parameters
    Public Const INPUT_DATA_TOO_LESS As Int32 = -3      ' too few parameters
    Public Const INPUT_DATA_ILLEGALITY As Int32 = -4    ' illegal parameter
    Public Const USB_WRITE_DATA_ERROR As Int32 = -5     ' USB write data error
    Public Const USB_READ_DATA_ERROR As Int32 = -6      ' USB read data error
    Public Const READ_NO_DATA As Int32 = -7             ' no data return when request data
    Public Const OPEN_DEVICE_FAILD As Int32 = -8        ' failed to open device
    Public Const CLOSE_DEVICE_FAILD As Int32 = -9       ' failed to close device
    Public Const EXECUTE_CMD_FAILD As Int32 = -10       ' the command failed to execute  
    Public Const SELECT_DEVICE_FAILD As Int32 = -11     ' failed to select device
    Public Const DEVICE_OPENED As Int32 = -12           ' device has open
    Public Const DEVICE_NOTOPEN As Int32 = -13          ' device not open
    Public Const BUFFER_OVERFLOW As Int32 = -14         ' buffer overflow
    Public Const DEVICE_NOTEXIST As Int32 = -15         ' device not exist
    Public Const LOAD_KERNELDLL As Int32 = -16          ' failed to load KernelDLL
    Public Const CMD_FAILED As Int32 = -17              ' failed to execute command
    Public Const BUFFER_CREATE As Int32 = -18           ' out of memory
    ' Define Adapter type
    Public Const VII_USBI2C As Int32 = 1
    ' The adapter type definition and data initialization definition
    Public Const VII_ADDR_7BIT As Byte = 7
    Public Const VII_ADDR_10BIT As Byte = 10
    Public Const VII_HCTL_MODE As Byte = 1
    Public Const VII_SCTL_MODE As Byte = 2
    Public Const VII_MASTER As Byte = 1
    Public Const VII_SLAVE As Byte = 0
    Public Const VII_SUB_ADDR_NONE As Byte = 0
    Public Const VII_SUB_ADDR_1BYTE As Byte = 1
    Public Const VII_SUB_ADDR_2BYTE As Byte = 2
    Public Const VII_SUB_ADDR_3BYTE As Byte = 3
    Public Const VII_SUB_ADDR_4BYTE As Byte = 4
    ' I2C data types initialization Define
    Public Structure VII_INIT_CONFIG
        Dim MasterMode As Byte      'Master-slave choice: 0-slave, 1-master
        Dim ControlMode As Byte     'Control mode: 0-Standard slave mode, 1-Standard mode, 2-GPIO mode
        Dim AddrType As Byte        '7-7bit mode, 10-10bit mode
        Dim SubAddrWidth As Byte    'Sub-Address width: value of 0~4, 0 means no Sub-Address mode
        Dim Addr As UInt16          'Device address in salve mode
        Dim ClockSpeed As UInt32    'Clock frequency(HZ)
    End Structure
    ' Adapter information data type
    Public Structure VII_BOARD_INFO
        <MarshalAs(UnmanagedType.ByValArray, SizeConst:=32)> _
        Dim ProductName As Byte()        'Product Name
        <MarshalAs(UnmanagedType.ByValArray, SizeConst:=4)> _
        Dim FirmwareVersion As Byte()    'Firmware Version
        <MarshalAs(UnmanagedType.ByValArray, SizeConst:=4)> _
        Dim HardwareVersion As Byte()    'Hardware Version
        <MarshalAs(UnmanagedType.ByValArray, SizeConst:=12)> _
        Dim SerialNumber As Byte()       'Serial Number
    End Structure
    ' I2C time parameter definition in GPIO mode(ms)
    Public Structure VII_TIME_CONFIG
        Dim tHD_STA As UInt16   ' Timing for start signal Keeping
        Dim tSU_STA As UInt16   ' Timing for start signal be established
        Dim tLOW As UInt16      ' Timing for clock low level
        Dim tHIGH As UInt16     ' Timing for clock high level
        Dim tSU_DAT As UInt16   ' Timing for data input be established
        Dim tSU_STO As UInt16   ' Timing for stop signal be established
        Dim tDH As UInt16       ' Timing for data output Keeping
        Dim tDH_DAT As UInt16   ' Timing for data input Keeping
        Dim tAA As UInt16       ' SCL lower to SDA output and response signal
        Dim tR As UInt16        ' Timing for SDA and SCL rising
        Dim tF As UInt16        ' Timing for SDA and SCL going down
        Dim tBuf As UInt16      ' Free timing of the bus until the new mission
        <MarshalAs(UnmanagedType.ByValArray, SizeConst:=4)> _
        Dim tACK As Byte()
        Dim tStart As UInt16
        Dim tStop As UInt16
    End Structure
    ' Function
    Declare Function VII_ScanDevice Lib "Ginkgo_Driver.dll" (ByVal NeedInit As Byte) As Int32
    Declare Function VII_OpenDevice Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal Reserved As Int32) As Int32
    Declare Function VII_CloseDevice Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32) As Int32
    Declare Function VII_ReadBoardInfo Lib "Ginkgo_Driver.dll" (ByVal DevIndex As Int32, ByRef pInfo As VII_BOARD_INFO) As Int32
    Declare Function VII_TimeConfig Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal I2CIndex As Int32, ByRef pTimeConfig As VII_TIME_CONFIG) As Int32
    Declare Function VII_InitI2C Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal I2CIndex As Int32, ByRef pInitConfig As VII_INIT_CONFIG) As Int32
    Declare Function VII_WriteBytes Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal I2CIndex As Int32, ByVal Addr As UInt16, ByVal SubAddr As UInt32, ByVal pWriteData() As Byte, ByVal Len As UInt16) As Int32
    Declare Function VII_ReadBytes Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal I2CIndex As Int32, ByVal Addr As UInt16, ByVal SubAddr As UInt32, ByVal pReadData() As Byte, ByVal Len As UInt16) As Int32
    Declare Function VII_SlaveReadBytes Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal I2CIndex As Int32, ByVal pReadData() As Byte, ByRef pLen As UInt16) As Int32
    Declare Function VII_SlaveWriteBytes Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal I2CIndex As Int32, ByVal pWriteData() As Byte, ByVal Len As UInt16) As Int32
    Declare Function VII_SetUserKey Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal pUserKey() As Byte) As Int32
    Declare Function VII_CheckUserKey Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal pUserKey() As Byte) As Int32
    Declare Function VII_SlaveWriteRemain Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal I2CIndex As Int32, ByRef pRemainBytes As UInt16) As Int32
End Module

File: Delphi/Ginkgo I2C/Delphi_USB_I2C_AT24C02_SCTL/ControlI2C.pas
unit ControlI2C;

interface

const
//Device type definition
VAI_USBADC	=	1;
VCI_USBCAN1	=	3;
VCI_USBCAN2	=	4;
VGI_USBGPIO	=	1;
VII_USBI2C	=	1;
VSI_USBSPI	=	2;
//USB-I2C适配器初始化数据定义
VII_ADDR_7BIT		=	7;	    // 7bit地址模式
VII_ADDR_10BIT	=	10;	    // 10bit地址模式
VII_HCTL_MODE		=	1;	    // 硬件控制
VII_SCTL_MODE		=	2;	    // 软件控制
VII_MASTER			=	1;	    // 主机
VII_SLAVE			  =	0;	    // 从机
VII_SUB_ADDR_NONE	  =	0;	// 无子地址
VII_SUB_ADDR_1BYTE	=	1;	// 1Byte子地址
VII_SUB_ADDR_2BYTE	=	2;	// 2Byte子地址
VII_SUB_ADDR_3BYTE	=	3;	// 3Byte子地址
VII_SUB_ADDR_4BYTE	=	4;	// 4Byte子地址
//错误码定义
ERR_SUCCESS					      =	(0);		// 没有错误
ERR_PARAMETER_NULL			  =	(-1);	  // 传入的指针为空指针
ERR_INPUT_DATA_TOO_MUCH		=	(-2);	  // 参数输入个数多余规定个数
ERR_INPUT_DATA_TOO_LESS		=	(-3);	  // 参数输入个数少余规定个数
ERR_INPUT_DATA_ILLEGALITY	=	(-4);	  // 参数传入格式和规定的不符合
ERR_USB_WRITE_DATA			  =	(-5);	  // USB写数据错误
ERR_USB_READ_DATA			    =	(-6);	  // USB读数据错误
ERR_READ_NO_DATA			    =	(-7);	  // 请求读数据时返回没有数据
ERR_OPEN_DEVICE				    =	(-8);	  // 打开设备失败
ERR_CLOSE_DEVICE			    =	(-9);	  // 关闭设备失败
ERR_EXECUTE_CMD				    =	(-10);	// 设备执行命令失败
ERR_SELECT_DEVICE			    =	(-11);	// 选择设备失败
ERR_DEVICE_OPENED			    =	(-12);	// 设备已经打开
ERR_DEVICE_NOTOPEN			  =	(-13);	// 设备没有打开
ERR_BUFFER_OVERFLOW			  =	(-14);	// 缓冲区溢出
ERR_DEVICE_NOTEXIST			  =	(-15);	// 此设备不存在
ERR_LOAD_KERNELDLL			  =	(-16);	// 装载动态库失败
ERR_CMD_FAILED				    =	(-17);	// 执行命令失败错误码
ERR_BUFFER_CREATE			    =	(-18);	// 内存不足
//USB-I2C初始化结构体
type
PVII_INIT_CONFIG = ^VII_INIT_CONFIG;
VII_INIT_CONFIG = record
  MasterMode:Byte;    //主从选择控制:0-从机，1-主机
  ControlMode:Byte;   //控制方式:1-硬件控制，2-软件控制
  AddrType:Byte;      //7-7bit模式，10-10bit模式
  SubAddrWidth:Byte;  //子地址宽度，0到4取值，0时表示无子地址模式
  Addr:Word;          //从机模式时候的设备地址
  ClockSpeed:Longword;//时钟频率:单位为HZ
end;
//软件I2C时间参数定义，时间单位为微秒
type
PVII_TIME_CONFIG = ^VII_TIME_CONFIG;
VII_TIME_CONFIG = record
  tHD_STA:Word;             //起始信号保持时间
  tSU_STA:Word;             //起始信号建立时间
  tLOW:Word;                //时钟低电平时间
  tHIGH:Word;               //时钟高电平时间
  tSU_DAT:Word;             //数据输入建立时间
  tSU_STO:Word;             //停止信号建立时间
  tDH:Word;                 //数据输出保持时间
  tDH_DAT:Word;             //数据输入保持时间
  tAA:Word;                 //SCL变低至SDA数据输出及应答信号
  tR:Word;                  //SDA及SCL上升时间
  tF:Word;                  //SDA及SCL下降时间
  tBuf:Word;                //新的发送开始前总线空闲时间
  tACK:Array[0..3] Of Byte;
  tStart:Word;
  tStop:Word;
end;
//函数声明
function VII_ScanDevice(NeedInit:Byte):Integer; stdcall;
function VII_OpenDevice(DevType,DevIndex,Reserved:Integer):Integer; stdcall;
function VII_CloseDevice(DevType,DevIndex:Integer):Integer; stdcall;
function VII_InitI2C(DevType, DevIndex, I2CIndex:Integer; pInitConfig:PVII_INIT_CONFIG):Integer; stdcall;
function VII_WriteBytes(DevType, DevIndex, I2CIndex:Integer;Addr:Word;SubAddr:Integer;pWriteData:PByte;Len:Word):Integer; stdcall;
function VII_ReadBytes(DevType, DevIndex, I2CIndex:Integer;Addr:Word;SubAddr:Integer;pReadData:PByte;Len:Word):Integer; stdcall;
function VII_TimeConfig(DevType, DevIndex, I2CIndex:Integer;pTimeConfig:PVII_TIME_CONFIG):Integer; stdcall;

implementation

function VII_ScanDevice;external 'Ginkgo_Driver.dll' name 'VII_ScanDevice';
function VII_OpenDevice;external 'Ginkgo_Driver.dll' name 'VII_OpenDevice';
function VII_CloseDevice;external 'Ginkgo_Driver.dll' name 'VII_CloseDevice';
function VII_InitI2C;external 'Ginkgo_Driver.dll' name 'VII_InitI2C';
function VII_WriteBytes;external 'Ginkgo_Driver.dll' name 'VII_WriteBytes';
function VII_ReadBytes;external 'Ginkgo_Driver.dll' name 'VII_ReadBytes';
function VII_TimeConfig;external 'Ginkgo_Driver.dll' name 'VII_TimeConfig';
end.
